pkg/virt-operator/creation/rbac/controller.go

```diff
@@ -147,10 +147,45 @@ func newControllerClusterRole() *rbacv1.ClusterRole {
 					"kubevirt.io",
 				},
 				Resources: []string{
-					"*",
+					"kubevirts",
 				},
 				Verbs: []string{
-					"*",
+					"get", "list", "watch",
+				},
+			},
+			{
+				APIGroups: []string{
+					"kubevirt.io",
+				},
+				Resources: []string{
+					"virtualmachines",
```

Comment:
With `make generate`, together with the suggested addition of verbs to various subjects, the following cluster-sync (= cluster-build + cluster-deploy) fails. Specifically, the cluster-deploy failed to deploy the 'kubevirts' instance, the 'make cluster-deply' step wait for 30 minutes and fail. 

I found just now, after repeated trying, following way to define the rules associated with `kubevirt.io` API Group would lead to the success of 'make cluster-deploy':

```
                        {
                                APIGroups: []string{
                                        "kubevirt.io",
                                },
                                Resources: []string{
                                        "*",
                                },
                                Verbs: []string{
                                        "get", "list", "watch", "patch", "update", "create",
                                },
                        },
                        {
                                APIGroups: []string{
                                        "kubevirt.io",
                                },
                                Resources: []string{
                                        "kubevirts",
                                        "virtualmachineinstancemigrations",
                                        // "virtualmachineinstancepresets",
                                        // "virtualmachineinstancereplicasets",
                                        // "virtualmachineinstances",
                                        "virtualmachines",
                                },
                                Verbs: []string{
                                        "get", "list", "watch", "patch", "update", "create", "delete",
                                },
                        },
```

The mystery for me right now is that in a successfully deployed functest environment, I see 6 CRD resources:
`vm, vmi, vmim, vmirs, vmipreset, kv`. However, when I try the following rule, 
```
                        {
                                APIGroups: []string{
                                        "kubevirt.io",
                                },
                                Resources: []string{
                                        "kubevirts",
                                        "virtualmachineinstancemigrations",
                                        "virtualmachineinstancepresets",
                                        "virtualmachineinstancereplicasets",
                                        "virtualmachineinstances",
                                        "virtualmachines",
                                },
                                Verbs: []string{
                                        "get", "list", "watch", "patch", "update", "create", "delete",
                                },
                        },
```
the `make cluster-deploy` fails. 

This led me to believe there are other resources in the API group for this rule specification would lead to success `make cluster-deploy`
```
                        {
                                APIGroups: []string{
                                        "kubevirt.io",
                                },
                                Resources: []string{
                                        "*",
                                },
                                Verbs: []string{
                                        "get", "list", "watch", "patch", "update", "create",
                                },
                        },
```

So my questions are:

- Are there other resources besides the listed 6 resources in the `kubevirt.io` API Group?

- The first pair of rules, including `create`, `update` for all resources in the API Group, with `delete` removed for some resource, and `deletecollection` removed for all resource, will this in the direction to address the requirement of the issue 3512 (https://github.com/kubevirt/kubevirt/issues/3512) ?

Reply:
One more question. On the additional changes induced by the `make generate`, I should not commit them before `push -f`, right? 

In addition, I was following this sequence for build and test. If you can give comment on this, it'd be great.
```
    make cluster-up
    make generate
    make cluster-build
    make cluster-deploy
    make functest       # w/ focus
    make cluster-down
```

Reply:
- Are there other resources besides the listed 6 resources in the kubevirt.io API Group?
No, at least in core. We have projects like `cdi` which have subgroup `cdi.kubevirt.io`. I am not sure if k8s actually recognize it as a subgroup or new group but that`s irrelevant.
- Yes, the goal is to restrict access for `virt-controller` as per my understanding. Basically what should be enough is to look at what verbs we are using under `pkg/virt-controller`.
- You can use `make cluster-sync` which does `make cluster-build` & `make cluster-deploy` for convinience. Your workflow is correct. You should run `make generate` after you make changes to code(Usually only to `types.go`, `schema.go`, `*.proto`, `crds.go` or `rbac`...).

FYI `pull-kubevirt-generate` runs `make generate-verify`(basicaly just check that nothing new gets generated.) This can fail on PR(usually not) even when it does not locally as every PR is rebased on master before testing.

Reply:
Thank you very much for the response. the 'make generate' suggestion were critical for me.